src/language-js/comments.js

```diff
@@ -929,6 +929,49 @@ function getCommentChildNodes(node, options) {
   ) {
     return [...(node.decorators || []), node.key, node.value.body];
   }
+
+  if (node.type === "ForStatement") {
+    const result = [];
+    for (const prop of ["init", "test", "update"]) {
+      result.push(
+        node[prop] ||
+          createCommentAnchorPseudoNode(
+            node,
+            prop,
+            result.length > 0
+              ? getNextNonSpaceNonCommentCharacterIndexWithStartIndex(
+                  text,
+                  indexOfSemicolon(text, locEnd(getLast(result))) + 1
+                )
+              : indexOfSemicolon(text, locStart(node))
+          )
+      );
+    }
+    result.push(node.body);
+    return result;
+  }
+}
+
+function indexOfSemicolon(text, startIdx) {
+  let nextIdx = startIdx;
+  for (;;) {
+    nextIdx = getNextNonSpaceNonCommentCharacterIndexWithStartIndex(
+      text,
+      nextIdx
+    );
+    if (nextIdx === false) {
+      /* istanbul ignore next */
+      return startIdx;
+    }
+    if (text.charAt(nextIdx) === ";") {
+      return nextIdx;
+    }
+    nextIdx++;
+  }
+}
+
+function createCommentAnchorPseudoNode(ref, marker, loc) {
+  return { type: "Prettier:CommentAnchor", ref, marker, range: [loc, loc] };
```

Comment:
Can we use `object` for this `marker`? I remeber you already use if for something in class, it should be more clear to understand what it is.

Reply:
It's a string label on a dangling comment needed to differentiate it from other dangling comments and to associate it with a specific place inside the enclosing node. Are you proposing not to use strings for this? I think strings work pretty well here: e.g. if the `init` part of a `for` statement is omitted, we might have dangling comments with the `init` marker there.

>  I remeber you already use if for something in class

For the same thing: to differentiate comments related to `extends` from comments related to `implements`, etc.

Reply:
I'm thinking about renaming `marker` to `anchor` as it seems to convey the purpose better, but let's do that in another PR.

Reply:
I did some renames, see the last commit.